### PR TITLE
Widgets: Add the ability to filter datasets by display tag

### PR DESCRIPTION
**Pull request by nadzpogi**
relates to xibosignage/xibo#3062

### CODE REVIEW

File: lib/Entity/DataSet.php [hunk 1]
```diff
@@ -492,6 +492,52 @@ public function getData($filterBy = [], $options = [])
             'connection' => 'default'
         ], $options);
 
+        // Fetch display tag value/s
+        if ($filter != '' && $displayId != 0) {
+            $displayTags = [];
+
+            // Define the regular expression to match [Tag:tagNameHere]
+            $pattern = '/\[Tag:(\w+)\]/';
```
Comment (dasgarner) [inline lib/Entity/DataSet.php:500]: Does this work with all allowed tag names? I.e. if I have a tag called `$storeNo` ?

File: lib/Entity/DataSet.php [hunk 2]
```diff
@@ -492,6 +492,52 @@ public function getData($filterBy = [], $options = [])
...
+            if (!empty($matches[1])) {
+                // Iterate through the matches and add them to the displayTags array
+                foreach ($matches[1] as $tag) {
+                    $displayTags[] = $tag;
+                }
+            }
+
+            // Loop through each tag and get the actual tag value from the database
+            foreach ($displayTags as $tag) {
+                $query = "
```
Comment (dasgarner) [inline lib/Entity/DataSet.php:515]: String should be enclosed in `'` instead of `"` please

File: lib/Entity/DataSet.php [hunk 3]
```diff
@@ -492,6 +492,52 @@ public function getData($filterBy = [], $options = [])
...
+                $params = [
+                    'displayId' => $displayId,
+                    'tag' => $tag
+                ];
+
+                // Execute the query
+                $results = $this->getStore()->select($query, $params);
+
+                // Check if results were found and get the tag value
+                $tagValue = !empty($results) ? $results[0]['tagValue'] : '';
```
Comment (dasgarner) [inline lib/Entity/DataSet.php:534]: I don't think this caters for the case where the display has the tag, but no value has been set?

**Review by dasgarner — CHANGES_REQUESTED** [review]
Looks good overall, but I think there are a few minor adjustments to make.

We also spoke about an extension to allow a default value to be provided along with the syntax, e.g. `[Tag:tagValue:defaultValue]` - we'd use the default value if the tag was not present on the display. Will you be adding that also?

File: lib/Entity/DataSet.php (hunk 2 above)
Comment (nadzpogi) [reply lib/Entity/DataSet.php]: Done!

File: lib/Entity/DataSet.php (hunk 1 above)
Comment (nadzpogi) [reply lib/Entity/DataSet.php]: I've replaced it now with a more generic regular expression to match `[Tag:...]` instead.

File: lib/Entity/DataSet.php [hunk 4]
```diff
@@ -494,47 +494,69 @@ public function getData($filterBy = [], $options = [])
...
-                $results = $this->getStore()->select($query, $params);
-
-                // Check if results were found and get the tag value
-                $tagValue = !empty($results) ? $results[0]['tagValue'] : '';
-
-                // Replace [Tag:anyTagNameHere] in the $filter with the actual tag value
-                $filter = preg_replace("/\[Tag:$tag\]/", $tagValue, $filter);
+                // Loop through each tag and get the actual tag value from the database
+                foreach ($displayTags as $tag) {
+                    $tagName = $tag['tagName'];
```
Comment (dasgarner) [inline lib/Entity/DataSet.php:527]: This is user input, so we need to load the parsed out $tag array into a sanitizer and ensure the values we've received are strings as expected. Have a think about where these values are used to see what I mean.

**Review by dasgarner — CHANGES_REQUESTED** [review]
Overall this is looking great - I have one final comment

File: lib/Entity/DataSet.php (hunk 4 above)
Comment (nadzpogi) [reply lib/Entity/DataSet.php]: Thanks for pointing that out. I've now loaded the $tag array into a sanitizer and made sure that the tagName and defaultValue that we're getting are strings.

**Review by dasgarner — APPROVED** [review]

**Comment by ifarzana** [comment]
TEST: /Library, /Layout END

**Pull request merged**